yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -96,6 +96,11 @@ private[yarn] class YarnAllocator(
   // was lost.
   private val pendingLossReasonRequests = new HashMap[String, mutable.Buffer[RpcCallContext]]
 
+  // Executor loss reason for explicitly released executors, it will be added when executor loss
```

Comment:
Given the race you described, this is not just for explicitly released executors, but for any executor whose loss reason hasn't been queried by the driver yet.